akka-projection-dynamodb/src/main/scala/akka/projection/dynamodb/internal/DynamoDBOffsetStore.scala

```diff
@@ -399,13 +408,15 @@ private[projection] class DynamoDBOffsetStore(
           records.filterNot(oldState.isDuplicate)
         else {
           // use last record for each pid
-          records
-            .groupBy(_.pid)
-            .valuesIterator
-            .collect {
-              case recordsByPid if !oldState.isDuplicate(recordsByPid.last) => recordsByPid.last
-            }
-            .toVector
+          val grouped = records.groupBy(_.pid)
+
+          if (grouped.size == records.size)
+            records.filterNot(oldState.isDuplicate)
+          else
+            grouped.valuesIterator
+              .map(_.maxBy(_.seqNr))
+              .filterNot(oldState.isDuplicate)
+              .toVector
```

Comment:
Why is this change needed? They are already by seqNr for each pid, and that order is preserved by `groupBy(_.pid)`. Similar was in https://github.com/akka/akka-projection/pull/1260 and then reverted https://github.com/akka/akka-projection/pull/1260/commits/04c6ce46adfc3b5aeec90bfbd9cd5aee1ae2fe7b

Reply:
I just keep coming back to not seeing where exactly the ordering is enforced: backtracking means it's not in the query source (for instance), so perhaps just a bit of paranoia here.